lib/tree.rb

```diff
@@ -16,40 +16,168 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n) must divide and conquer based on if new key is > || <
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      return @root = new_node
+    end
+
+    current = @root
+    add_helper(current, new_node)
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(root, add_node)
```

Comment:
I like the helpers you created.  Just a note, you could even put these helpers in the `TreeNode` class.